subsys/net/l2/ppp/fsm.c

```diff
@@ -0,0 +1,1088 @@
+		ppp_state_str(fsm->state), fsm->state);
+
+	switch (fsm->state) {
+	case PPP_CLOSING:
+		new_state = PPP_CLOSED;
+		goto stopped;
+
+	case PPP_OPENED:
+		if (fsm->cb.down) {
+			fsm->cb.down(fsm);
+		}
+
+		fsm_send_configure_req(fsm, false);
+		ppp_change_state(fsm, PPP_REQUEST_SENT);
+		break;
+
+	case PPP_STOPPING:
+		new_state = PPP_STOPPED;
+		goto stopped;
+
+	case PPP_ACK_RECEIVED:
+		ppp_change_state(fsm, PPP_REQUEST_SENT);
+		break;
+
+	default:
+		NET_DBG("[%s/%p] %s state %s (%d)", fsm->name, fsm, "Invalid",
+			ppp_state_str(fsm->state), fsm->state);
+		return NET_DROP;
+	}
+
+	return NET_OK;
+
+stopped:
+	k_delayed_work_cancel(&fsm->timer);
+	ppp_change_state(fsm, new_state);
+
+	if (fsm->cb.finished) {
+		fsm->cb.finished(fsm);
+	}
+
+	return NET_OK;
+}
+
+static enum net_verdict fsm_recv_code_rej(struct ppp_fsm *fsm,
+					  struct net_pkt *pkt)
+{
+	u8_t code, id;
+	int ret;
+
+	NET_DBG("[%s/%p] Current state %s (%d)", fsm->name, fsm,
+		ppp_state_str(fsm->state), fsm->state);
+
+	ret = net_pkt_read_u8(pkt, &code);
+	if (ret < 0) {
+		return NET_DROP;
+	}
+
+	ret = net_pkt_read_u8(pkt, &id);
+	if (ret < 0) {
+		return NET_DROP;
+	}
+
+	NET_DBG("[%s/%p] Received Code-Rej code %d id %d", fsm->name, fsm,
+		code, id);
+
+	if (fsm->state == PPP_ACK_RECEIVED) {
+		ppp_change_state(fsm, PPP_REQUEST_SENT);
+	}
+
+	return NET_OK;
+}
+
+void ppp_fsm_proto_reject(struct ppp_fsm *fsm)
+{
+	NET_DBG("[%s/%p] Current state %s (%d)", fsm->name, fsm,
+		ppp_state_str(fsm->state), fsm->state);
+
+	switch (fsm->state) {
+	case PPP_ACK_RECEIVED:
+	case PPP_ACK_SENT:
+	case PPP_STOPPING:
+	case PPP_REQUEST_SENT:
+		k_delayed_work_cancel(&fsm->timer);
+		ppp_change_state(fsm, PPP_STOPPED);
+		if (fsm->cb.finished) {
+			fsm->cb.finished(fsm);
+		}
+
+		break;
+
+	case PPP_CLOSED:
+		ppp_change_state(fsm, PPP_CLOSED);
+		if (fsm->cb.finished) {
+			fsm->cb.finished(fsm);
+		}
+
+		break;
+
+	case PPP_CLOSING:
+		k_delayed_work_cancel(&fsm->timer);
+		ppp_change_state(fsm, PPP_CLOSED);
+		if (fsm->cb.finished) {
+			fsm->cb.finished(fsm);
+		}
+
+		break;
+
+	case PPP_OPENED:
+		terminate(fsm, PPP_STOPPING);
+		break;
+
+	case PPP_STOPPED:
+		ppp_change_state(fsm, PPP_STOPPED);
+		if (fsm->cb.finished) {
+			fsm->cb.finished(fsm);
+		}
+
+		break;
+
+	default:
+		NET_DBG("[%s/%p] %s state %s (%d)", fsm->name, fsm, "Invalid",
+			ppp_state_str(fsm->state), fsm->state);
+		break;
+	}
+}
+
+enum net_verdict ppp_fsm_input(struct ppp_fsm *fsm, u16_t proto,
+			       struct net_pkt *pkt)
+{
+	u8_t code, id;
+	u16_t length;
+	int ret;
+
+	ret = net_pkt_read_u8(pkt, &code);
+	if (ret < 0) {
+		NET_DBG("[%s/%p] Cannot read %s (pkt len %zd)",
+			fsm->name, fsm, "code", net_pkt_get_len(pkt));
+		return NET_DROP;
+	}
+
+	ret = net_pkt_read_u8(pkt, &id);
+	if (ret < 0) {
+		NET_DBG("[%s/%p] Cannot read %s (pkt len %zd)",
+			fsm->name, fsm, "id", net_pkt_get_len(pkt));
+		return NET_DROP;
+	}
+
+	ret = net_pkt_read_be16(pkt, &length);
```

Comment:
Yes, it could be done like that too. The PPP protocol does not specify that kind of struct in the RFC so I did it by reading the values individually. I do not want to start to change that at this point, but can be tweaked later.